### PR TITLE
Add `ReferenceStorage` for manual allocation of references

### PR DESCRIPTION
**Pull request by HertzDevil**
This adds the `Instance` type mentioned in #13481. The naming is not final.

### CODE REVIEW

File: src/compiler/crystal/types.cr [hunk 1]
```diff
@@ -2642,6 +2642,32 @@ module Crystal
...
 
+  # The non-instantiated ReferenceStorage(T) type.
+  class GenericReferenceStorageType < GenericClassType
+    @struct = true
+
+    def new_generic_instance(program, generic_type, type_vars)
+      t = type_vars["T"].type
+
+      unless t.is_a?(TypeParameter) || (t.class? && !t.struct?)
+        raise TypeException.new "can't instantiate ReferenceStorage(T) with T = #{t} (T must be a reference type)"
```
Comment (Sija) [inline src/compiler/crystal/types.cr:2653]:
```suggestion
        raise TypeException.new "Can't instantiate ReferenceStorage(T) with T = #{t} (T must be a reference type)"
```

**Review by straight-shoota — APPROVED** [review]
LGTM.

As mentioned in the OP, naming is still kind of preliminary. But in absence of any really clever idea, we can just take `ReferenceStorage`.
In case there is something inherently wrong about it, we can still change it as it's marked as experimental.

**Comment by HertzDevil** [comment]
What about the minimal implementations for `Object`'s abstract defs? Do you think they make sense?

**Comment by straight-shoota** [comment]
Not sure... Maybe we could leave them out for now and add them later if we notice they're missing? It's easy to patch them in ad-hoc, so this isn't holding up anything.

I suppose instances of `ReferenceStorage` wouldn't really be used on their own. The most typical use case we have been thinking of (so far):
```cr
just_some_variable_to_put_this_on_the_stack = uninitialized ReferenceStorage(Foo)
foo = Foo.pre_initialize(pointerof(just_some_variable_to_put_this_on_the_stack))
```
And `just_some_variable_to_put_this_on_the_stack` is never used again (unless you want to allocate a something else in that storage, but that would be the same usage as the second line here).

Probably, this will be abstracted away in a macro for convenience, further reducing direct usage of `ReferenceStorage`.

**Comment by HertzDevil** [comment]
The problem is they cannot be left out at all because the type inherits directly from `Value` rather than `Struct`. (I guess you could leave out `#==` because `Value` has a catch-all def that returns false)

**Comment by straight-shoota** [comment]
Sounds good. Let's keep it then 👍

**Comment by straight-shoota** [comment]
This was eventually reverted in https://github.com/crystal-lang/crystal/pull/14207 for 1.11.1 and an alternative impementation is available at #14204

**Pull request merged**